### PR TITLE
Docs: Add "React peer dependency is no longer required" section to addon migration guide

### PR DESCRIPTION
**Pull request by vanessayuenn**
## What I did

Added the bit about addon now can remove their react peer dependency.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

No need!

### Documentation

- [x] Add or update documentation reflecting your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: docs/addons/addon-migration-guide.md [hunk 1]
```diff
@@ -34,6 +34,15 @@ Please upgrade your addon to Node.js 18, as support for Node.js 16 has ended.
 
 UI injected into panels, tools, etc. by addons is now rendered with React 18. Also note that the `key` prop is no longer passed to the render function.
 
+### React peer dependency is no longer required
```
Comment (jonniebigodes) [inline docs/addons/addon-migration-guide.md:37]: @vanessayuenn, when you have a moment, can you adjust this section to include a mention for users who are using an outdated version of the addon kit that was pre-tsup to reach out to us via our communication channels (e.g., GitHub discussions) as we don't have a direct path for such cases?

**Review by jonniebigodes — CHANGES_REQUESTED** [review]
@vanessayuenn appreciate you taking the time to put together this pull request and help improve the documentation 🙏 ! I left a small item for you to look into when you have a moment. 

Let me know of any questions you may have so we can address them and make this available to our users.

Have a great day.

Stay safe

File: docs/addons/addon-migration-guide.md [hunk 2]
```diff
@@ -34,6 +34,17 @@ Please upgrade your addon to Node.js 18, as support for Node.js 16 has ended.
...
+
+To remove your addon's peer dependency on React, and reduce its install size, do the following:
+
+1. Move `react`, `react-dom` and the globalized Storybook packages from `peerDependencies` to `devDependencies`
+2. Add the list of globalized packages to the `externals` property in the `tsup` configuration, to ensure they are not part of the bundle.
+
+For an example, see [the updates we've made to the addon-kit](https://github.com/storybookjs/addon-kit/pull/60/files#diff-8fed899bdbc24789a7bb4973574e624ed6207c6ce572338bc3c3e117672b2a20). These changes are optional but recommended.
+
+Note: This assumes your addon is using tsup for bundling. If your addon was built with an older version of addon-kit that uses Babel for bundling, you have to first switch to tsup. For guidance, explore these [older changes implemented in the addon-kit](https://github.com/storybookjs/addon-kit/pull/45/files).
```
Comment (vanessayuenn) [inline docs/addons/addon-migration-guide.md:46]: @jonniebigodes I added this line here for pre-tsup addon authors. Does this address your concern?

File: docs/addons/addon-migration-guide.md (hunk 1 above)
Comment (vanessayuenn) [reply docs/addons/addon-migration-guide.md:37]: see this [comment](https://github.com/storybookjs/storybook/pull/25826/files#r1488518864)

File: docs/addons/addon-migration-guide.md (hunk 2 above)
Comment (jonniebigodes) [reply docs/addons/addon-migration-guide.md:46]: It works for me, but we can do a bit better and wrap the "Note:" in a callout, turning this into:

```md
To remove your addon's peer dependency on React, and reduce its install size, do the following:

1. Move `react`, `react-dom` and the globalized Storybook packages from `peerDependencies` to `devDependencies`
2. Add the list of globalized packages to the `externals` property in the `tsup` configuration, to ensure they are not part of the bundle.

For an example, see [the updates we've made to the addon-kit](https://github.com/storybookjs/addon-kit/pull/60/files#diff-8fed899bdbc24789a7bb4973574e624ed6207c6ce572338bc3c3e117672b2a20). These changes are optional but recommended.

<Callout variant="info">

This assumes your addon is using tsup for bundling. If your addon was built with an older version of addon-kit that uses Babel for bundling, you have to first switch to tsup. For guidance, explore these [older changes implemented in the addon-kit](https://github.com/storybookjs/addon-kit/pull/45/files).

</Callout>
```

Comment (vanessayuenn) [reply docs/addons/addon-migration-guide.md:46]:
```suggestion
<Callout variant="info"> 

This assumes your addon is using tsup for bundling. If your addon was built with an older version of addon-kit that uses Babel for bundling, you have to first switch to tsup. For guidance, explore these [older changes implemented in the addon-kit](https://github.com/storybookjs/addon-kit/pull/45/files). 

</Callout>
```

**Comment by JReinhold** [comment]
@vanessayuenn @jonniebigodes I added a section referencing the addon kit changes as a migration example.

File: docs/addons/addon-migration-guide.md [hunk 3]
```diff
@@ -58,6 +73,16 @@ As a workaround, update your documentation to tell users to opt-in to Babel supp
...
+
+[This complete diff view](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f) shows everything changed between 7.0 and 8.0, but not all of it is relevant for your migration. You can specifically focus on:
+
+- [`package.json`](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519): Updating dependencies, moving to ESM, updating entries
+- [tsup.config.ts](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f#diff-8fed899bdbc24789a7bb4973574e624ed6207c6ce572338bc3c3e117672b2a20): Updating bundling strategy to take advantage of Storybook globals.
+- [`.storybook/local-preset.js`](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f#diff-390b53ea479b1ceffcbf31944f644ee23aa9f337b75a8a0ffd815bed50d376cb): Supporting the ESM migration.
```
Comment (jonniebigodes) [inline docs/addons/addon-migration-guide.md:84]: @JReinhold, When you have a moment, please address this to the suggestion below, as this entire section still required a bit of work before this was merged.

```suggestion
## Migration Example

The Addon Kit [repository](https://github.com/storybookjs/addon-kit) has already been updated to support Storybook 8.0, and you can use it as a reference for your migration. You'll see the changes mentioned in this guide, including ESM support via the `type: module` configuration. As an addon maintainer, we encourage you to update your addon to include them. It simplifies the setup and makes it easier for users to use your addon with the latest version of Storybook. If you choose to follow along with the ESM migration, we've prepared an abbreviated list of changes below.

- [`package.json`](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519) for dependencies management, ESM support and updates to the addon's entry points.
- [`tsup.config.ts`](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f#diff-8fed899bdbc24789a7bb4973574e624ed6207c6ce572338bc3c3e117672b2a2) for bundling changes, factoring in Storybook's globals.
- [`.storybook/local-preset.js`](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f#diff-390b53ea479b1ceffcbf31944f644ee23aa9f337b75a8a0ffd815bed50d376cb) to support the ESM migration.

<Callout variant="info" icon="💡">

If you migrate your addon to support ESM, you can safely remove any `manager.js`, `preview.js`, and `preset.js` files from the addon's root directory. For a complete overview of the changes applied to the Addon Kit to fully support Storybook 8.0, see the following [diff view](https://github.com/storybookjs/addon-kit/compare/79282986..cf0875f).

</Callout>
```

**Comment by jonniebigodes** [comment]
@vanessayuenn I left a suggestion for @JReinhold to look into for the migration example. However, the pull request that superseded this one was approved and merged without me having a chance to look into it, and this introduced some documentation issues that could have been easily prevented. Once this is merged, I'll do a proper vetting and polish this to get it into a good state and avoid any follow-up pull requests further down the line.

**Comment by vanessayuenn** [comment]
@jonniebigodes i will be merging this for now as this PR contains crucial information for addon authors to migrate. Please do the polishing separately though! Appreciate your help here ❤️ .

**Pull request merged**